File: leetcode/1019-next-greater-node-in-linked-list/Solution.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nextLargerNodes(ListNode head) {
        List<Integer> values = new ArrayList<>();
        for (ListNode node = head; node != null; node = node.next) {
            values.add(node.val);
        }

        int n = values.size();
        int[] result = new int[n];
        Deque<Integer> stack = new ArrayDeque<>(); // indices awaiting their next greater value

        for (int i = 0; i < n; i++) {
            int val = values.get(i);
            while (!stack.isEmpty() && values.get(stack.peek()) < val) {
                result[stack.pop()] = val;
            }
            stack.push(i);
        }

        return result;
    }
}

File: leetcode/1019-next-greater-node-in-linked-list/solution.cpp
#include <vector>
using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        vector<int> values;
        for (ListNode* node = head; node; node = node->next) {
            values.push_back(node->val);
        }

        int n = values.size();
        vector<int> result(n, 0);
        vector<int> stack; // indices awaiting their next greater value

        for (int i = 0; i < n; i++) {
            while (!stack.empty() && values[stack.back()] < values[i]) {
                result[stack.back()] = values[i];
                stack.pop_back();
            }
            stack.push_back(i);
        }

        return result;
    }
};
